### PR TITLE
Validate unsearchable

### PR DESCRIPTION
**Pull request by yucohen**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/42115

## Description
The `unsearchable` field in incident, indicator and genetic fields should be true.

## Must have
- [ ] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

File: demisto_sdk/commands/common/hook_validations/generic_field.py [hunk 1]
```diff
@@ -55,3 +58,18 @@ def is_valid_id_prefix(self):
             return False
 
         return True
+
+    def is_valid_unsearchable_field(self):
+        # type: () -> bool
+        """Validate that the unsearchable field is true
+        (relevant to incident_field and generic_field)
+        Returns:
+            bool. Whether the file's unsearchable field is se to true.
```
Comment (reutshal) [inline demisto_sdk/commands/common/hook_validations/generic_field.py:67]:
```suggestion
            bool. Whether the file's unsearchable key is set to true.
```

File: demisto_sdk/commands/common/hook_validations/incident_field.py [hunk 2]
```diff
@@ -426,3 +428,18 @@ def is_valid_incident_field_name_prefix(self):
                 return False
 
         return True
+
+    def is_valid_unsearchable_field(self):
+        # type: () -> bool
+        """Validate that the unsearchable field is true
+        (relevant to incident_field and generic_field)
+        Returns:
+            bool. Whether the file's unsearchable field is se to true.
```
Comment (reutshal) [inline demisto_sdk/commands/common/hook_validations/incident_field.py:437]: remember to change to "key"

File: demisto_sdk/commands/validate/tests/validators_test.py [hunk 3]
```diff
@@ -1157,6 +1159,27 @@ def test_ignore_test_doc_non_pack_file_non_pack(self, file_path: str):
         validate_manager = ValidateManager(check_is_unskipped=False)
         assert validate_manager.ignore_test_doc_non_pack_file(file_path)
 
+    @pytest.mark.parametrize('answer, unsearchable, validator', [(True, True, IncidentFieldValidator),
```
Comment (reutshal) [inline demisto_sdk/commands/validate/tests/validators_test.py:1162]:
```suggestion
    @pytest.mark.parametrize('expected_result, unsearchable, validator', [(True, True, IncidentFieldValidator),
```

I think this name is better for your test

File: demisto_sdk/tests/test_files/validate_integration_test_valid_types.py [hunk 4]
```diff
@@ -182,7 +182,7 @@
     "systemAssociatedTypes": None,
     "associatedToAll": False,
     "unmapped": False,
-    "unsearchable": False,
+    "unsearchable": True,
```
Comment (reutshal) [inline demisto_sdk/tests/test_files/validate_integration_test_valid_types.py:185]: Some of the values are lowercased and some uppercased. Is there a reason?
If not, please let's be consistent.

**Review by reutshal — CHANGES_REQUESTED** [review]
Looks good, please see my comments

File: demisto_sdk/commands/common/errors.py [hunk 5]
```diff
@@ -421,13 +423,13 @@ def indicator_field_type_grid_minimal_version(fromversion):
 
     @staticmethod
     @error_code_decorator
-    def fields_unsearchable_should_be_true_if():
-        return 'The unsearchable field in indicator and incident field should be set to true.'
+    def unsearchable_key_should_be_true_if():
+        return 'The unsearchable key in indicator and incident field should be set to true.'
```
Comment (reutshal) [inline demisto_sdk/commands/common/errors.py:427]:
```suggestion
        return 'The unsearchable key in indicator and incident fields should be set to true.'
```

File: demisto_sdk/commands/common/errors.py [hunk 6]
```diff
@@ -421,13 +423,13 @@ def indicator_field_type_grid_minimal_version(fromversion):
...
 
     @staticmethod
     @error_code_decorator
-    def fields_unsearchable_should_be_true_gf():
-        return 'The unsearchable field in generic field should be set to true.'
+    def unsearchable_key_should_be_true_gf():
+        return 'The unsearchable key in generic field should be set to true.'
```
Comment (reutshal) [inline demisto_sdk/commands/common/errors.py:432]:
```suggestion
        return 'The unsearchable key in a generic field should be set to true.'
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,6 @@
 # Changelog
 * Fixed an issue where the **update-release-notes** command crashes with `--help` flag.
+* Added a validation to the **validate** command that verifies that `unsearchable` field in incident, indicator and generic fields is set to true.
```
Comment (reutshal) [inline CHANGELOG.md:3]:
```suggestion
* Added validation to the **validate** command that verifies the `unsearchable` key in incident, indicator and generic fields is set to true.
```

File: demisto_sdk/commands/common/errors.py [hunk 8]
```diff
@@ -13,6 +13,9 @@
 ALLOWED_IGNORE_ERRORS = [
     'BA101', 'BA106', 'BA108', 'BA109', 'BA110', 'BA111', 'BA112', 'BA113',
     'DS107',
+    'GF102',
+    'IF100', 'IF106', 'IF115',
+    'IN109', 'IN110', 'IN122', 'IN126', 'IN128', 'IN135', 'IN136', 'IN139', 'IN144',
```
Comment (reutshal) [inline demisto_sdk/commands/common/errors.py:18]: Why was this line added?

File: demisto_sdk/commands/validate/tests/validators_test.py [hunk 9]
```diff
@@ -1160,6 +1162,44 @@ def test_ignore_test_doc_non_pack_file_non_pack(self, file_path: str):
         validate_manager = ValidateManager(check_is_unskipped=False)
         assert validate_manager.ignore_test_doc_non_pack_file(file_path)
 
+    @pytest.mark.parametrize('expected_result, unsearchable', [(True, True),
+                                                               (False, False)]
+                             )
+    def test_is_valid_unsearchable_key_if(self, pack: Pack, expected_result, unsearchable):
```
Comment (reutshal) [inline demisto_sdk/commands/validate/tests/validators_test.py:1168]:
```suggestion
    def test_is_valid_unsearchable_key_incident_field(self, pack: Pack, expected_result, unsearchable):
```

File: demisto_sdk/commands/validate/tests/validators_test.py [hunk 10]
```diff
@@ -1160,6 +1162,44 @@ def test_ignore_test_doc_non_pack_file_non_pack(self, file_path: str):
...
+        incident_field = pack.create_incident_field('MyIncidentField')
+        incident_field.update({"unsearchable": unsearchable})
+        structure = StructureValidator(incident_field.path)
+        res_validator = IncidentFieldValidator(structure)
+        assert res_validator.is_valid_unsearchable_key() is expected_result
+
+    @pytest.mark.parametrize('expected_result, unsearchable', [(True, True),
+                                                               (False, False)]
+                             )
+    def test_is_valid_unsearchable_key_gf(self, pack: Pack, expected_result, unsearchable):
```
Comment (reutshal) [inline demisto_sdk/commands/validate/tests/validators_test.py:1187]:
```suggestion
    def test_is_valid_unsearchable_key_generic_field(self, pack: Pack, expected_result, unsearchable):
```

**Review by reutshal — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/common/errors.py (hunk 8 above)
Comment (yucohen) [reply demisto_sdk/commands/common/errors.py:18]: merge conflicts 🤦🏼‍♀️

**Review by reutshal — APPROVED** [review]

File: demisto_sdk/commands/common/hook_validations/generic_field.py [hunk 11]
```diff
@@ -55,3 +58,17 @@ def is_valid_id_prefix(self):
...
 
         return True
+
+    def is_valid_unsearchable_key(self):
+        # type: () -> bool
+        """Validate that the unsearchable key is set to true
+        Returns:
+            bool. Whether the file's unsearchable key is set to true.
+        """
+        indicator_field_unsearchable = self.current_file.get('unsearchable', True)
```
Comment (tomneeman151293) [inline demisto_sdk/commands/common/hook_validations/generic_field.py:68]:
```suggestion
        generic_field_unsearchable = self.current_file.get('unsearchable', True)
```

File: demisto_sdk/commands/common/hook_validations/generic_field.py [hunk 12]
```diff
@@ -55,3 +58,17 @@ def is_valid_id_prefix(self):
...
+        if indicator_field_unsearchable:
```
Comment (tomneeman151293) [inline demisto_sdk/commands/common/hook_validations/generic_field.py:69]:
```suggestion
        if generic_field_unsearchable:
```

File: demisto_sdk/commands/validate/validate_manager.py [hunk 13]
```diff
@@ -854,17 +857,20 @@ def validate_report(self, structure_validator, pack_error_ignore_list):
...
                                                           print_as_warnings=self.print_ignored_errors,
                                                           json_file_path=self.json_file_path)
         if is_modified and self.is_backward_check:
             return all([incident_field_validator.is_valid_file(validate_rn=False, is_new_file=not is_modified,
-                                                               use_git=self.use_git),
-                        incident_field_validator.is_backward_compatible()])
+                                                               use_git=self.use_git,
+                                                               is_added_file=is_added_file),
+                        incident_field_validator.is_backward_compatible(),
+                        incident_field_validator.is_valid_unsearchable_key()])
```
Comment (tomneeman151293) [inline demisto_sdk/commands/validate/validate_manager.py:869]: why is this needed here? shouldn't it already be executed in is_valid_file?

**Review by tomneeman151293 — CHANGES_REQUESTED** [review]
Nice. Please also add the following scenarios to is_valid_file  test for both generic field and indicator field
1. added_file is False, unsearchable is False
2. added_file is False, unsearchable is True
3. added_file is True, unsearchable is False
4. added_file is True, unsearchable is True

**Comment by tomneeman151293** [comment]
Also - need to make sure running format on an incident/indicator field changes this field to true (and add relevant tests for it)

File: demisto_sdk/commands/format/update_generic.py [hunk 14]
```diff
@@ -365,3 +365,9 @@ def initiate_file_validator(self) -> int:
 
             else:
                 return SUCCESS_RETURN_CODE
+
+    def update_unsearchable_key(self):
+        """Changes 'unsearchable' key of a generic field object to default."""
+        if self.verbose:
+            click.echo(f'Setting unsearchable to default: {UNSEARCHABLE_KEY_DEFAULT_VALUE}')
+        self.data['unsearchable'] = UNSEARCHABLE_KEY_DEFAULT_VALUE
```
Comment (tomneeman151293) [inline demisto_sdk/commands/format/update_generic.py:373]: Not needed after the requested change is done

File: demisto_sdk/commands/format/update_genericfield.py [hunk 15]
```diff
@@ -33,6 +33,7 @@ def run_format(self) -> int:
             click.secho(f'\n================= Updating file {self.source_file} =================', fg='bright_blue')
             super().update_json()
             self.set_default_values_as_needed()
+            self.update_unsearchable_key()
```
Comment (tomneeman151293) [inline demisto_sdk/commands/format/update_genericfield.py:36]: Not needed after the requested change is done

File: demisto_sdk/commands/format/update_incidentfields.py [hunk 16]
```diff
@@ -32,6 +32,7 @@ def run_format(self) -> int:
             click.secho(f'\n================= Updating file {self.source_file} =================', fg='bright_blue')
             super().update_json()
             self.set_default_values_as_needed()
+            self.update_unsearchable_key()
```
Comment (tomneeman151293) [inline demisto_sdk/commands/format/update_incidentfields.py:35]: Not needed after the requested change is done

File: demisto_sdk/commands/format/update_indicatorfields.py [hunk 17]
```diff
@@ -31,6 +31,7 @@ def run_format(self) -> int:
             click.secho(f'\n================= Updating file {self.source_file} =================', fg='bright_blue')
             super().update_json()
             self.set_default_values_as_needed()
+            self.update_unsearchable_key()
```
Comment (tomneeman151293) [inline demisto_sdk/commands/format/update_indicatorfields.py:34]: Not needed after the requested change is done

File: demisto_sdk/commands/format/format_constants.py [hunk 18]
```diff
@@ -14,6 +14,7 @@
 SKIP_RETURN_CODE = 2
 SKIP_VALIDATE_PY_RETURN_CODE = 3
 GENERIC_FIELD_DEFAULT_GROUP = 4
+UNSEARCHABLE_KEY_DEFAULT_VALUE = True
```
Comment (tomneeman151293) [inline demisto_sdk/commands/format/format_constants.py:17]: Can remove this

File: demisto_sdk/commands/common/hook_validations/generic_field.py [hunk 19]
```diff
@@ -55,3 +58,17 @@ def is_valid_id_prefix(self):
...
+        # type: () -> bool
+        """Validate that the unsearchable key is set to true
+        Returns:
+            bool. Whether the file's unsearchable key is set to true.
+        """
+        generic_field_unsearchable = self.current_file.get('unsearchable', True)
+        if generic_field_unsearchable:
+            return True
+        error_message, error_code = Errors.unsearchable_key_should_be_true_generic_field()
+        if self.handle_error(error_message, error_code, file_path=self.file_path):
```
Comment (tomneeman151293) [inline demisto_sdk/commands/common/hook_validations/generic_field.py:72]:
```suggestion
        if self.handle_error(error_message, error_code, file_path=self.file_path, suggested_fix=Errors.suggest_fix(self.file_path)):
```

File: demisto_sdk/commands/common/hook_validations/incident_field.py [hunk 20]
```diff
@@ -426,3 +428,17 @@ def is_valid_incident_field_name_prefix(self):
...
+        # type: () -> bool
+        """Validate that the unsearchable key is true
+        Returns:
+            bool. Whether the file's unsearchable key is set to true.
+        """
+        unsearchable = self.current_file.get('unsearchable', True)
+        if unsearchable:
+            return True
+        error_message, error_code = Errors.unsearchable_key_should_be_true_incident_field()
+        if self.handle_error(error_message, error_code, file_path=self.file_path):
```
Comment (tomneeman151293) [inline demisto_sdk/commands/common/hook_validations/incident_field.py:442]:
```suggestion
        if self.handle_error(error_message, error_code, file_path=self.file_path, suggested_fix=Errors.suggest_fix(self.file_path)):
```

**Review by tomneeman151293 — CHANGES_REQUESTED** [review]
Very nice! One last small note tho

There is no need to add a specific function just for unsearchable.
Please see `ARGUMENTS_DEFAULT_VALUES` in fromat_constants.
I think adding a pair of
'unsearchable': (True, ['IncidentFieldJSONFormat', 'IndicatorFieldJSONFormat', 'GenericFieldJSONFormat'])
Will make the format change unsearchable to true without the needing of unsearchable func, see `set_default_values_as_needed` in update_generic_json.py.
 so you won't need to modify any format/update_incident/generic/indicator field.py files.

Also - I might have missed but did you add tests to the format changes?

**Review by tomneeman151293 — APPROVED** [review]
Great Work!

**Comment by tomneeman151293** [comment]
Please re-trigger run after this PR is merged
https://github.com/demisto/content/pull/15798
it will fix your validate files failure (which isn't related to you)

**Pull request merged**